lxd/container_lxc.go

```diff
@@ -1889,25 +1889,112 @@ func (c *containerLXC) deviceStart(deviceName string, rawConfig map[string]strin
+		// Shift device file ownership if needed before mounting into container.
+		// This needs to be done whether or not container is running.
+		if len(runConfig.Mounts) > 0 {
+			err := c.deviceShiftMounts(runConfig.Mounts)
 			if err != nil {
 				return nil, err
 			}
 		}
 
-		err = c.runHooks(runConfig.PostHooks)
-		if err != nil {
-			return nil, err
+		// If container is running and then live attach device.
+		if isRunning {
+			// Attach mounts if requested.
+			if len(runConfig.Mounts) > 0 {
+				err = c.deviceAttachMounts(configCopy, runConfig.Mounts)
+				if err != nil {
+					return nil, err
+				}
+			}
+
+			// Add cgroup rules if requested.
+			if len(runConfig.CGroups) > 0 {
+				err = c.deviceAddCgroupRules(configCopy, runConfig.CGroups)
+				if err != nil {
+					return nil, err
+				}
+			}
+
+			// Attach network interface if requested.
+			if len(runConfig.NetworkInterface) > 0 {
+				err = c.deviceAttachNIC(configCopy, runConfig.NetworkInterface)
+				if err != nil {
+					return nil, err
+				}
+			}
+
+			err = c.runHooks(runConfig.PostHooks)
+			if err != nil {
+				return nil, err
+			}
 		}
 	}
 
 	return runConfig, nil
 }
 
+// deviceShiftMounts shift device mount files to active idmap if needed.
+func (c *containerLXC) deviceShiftMounts(mounts []device.MountEntryItem) error {
+	idmapSet, err := c.CurrentIdmap()
+	if err != nil {
+		return fmt.Errorf("Failed to get idmap for device: %s", err)
+	}
+
+	// If there is an idmap being applied and LXD not running in a user namespace then shift the
+	// device files before they are mounted.
+	if idmapSet != nil && !c.state.OS.RunningInUserNS {
+		for _, mount := range mounts {
+			if err := idmapSet.ShiftFile(mount.DevPath); err != nil {
+				// uidshift failing is weird, but not a big problem. Log and proceed.
+				logger.Debugf("Failed to uidshift device %s: %s\n", mount.DevPath, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// deviceAttachMounts live attaches mounts to a container.
+func (c *containerLXC) deviceAttachMounts(configCopy map[string]string, mounts []device.MountEntryItem) error {
+	for _, mount := range mounts {
+		flags := 0
+
+		// Convert options into flags.
+		for _, opt := range mount.Opts {
+			if opt == "bind" {
+				flags |= unix.MS_BIND
+			}
+		}
+
+		// Mount it into the container.
+		err := c.insertMount(mount.DevPath, mount.TargetPath, mount.FSType, flags, mount.Shift)
+		if err != nil {
+			return fmt.Errorf("Failed to add mount for device: %s", err)
+		}
+	}
+
+	return nil
+}
+
+// deviceAddCgroupRules live adds cgroup rules to a container.
+func (c *containerLXC) deviceAddCgroupRules(configCopy map[string]string, cgroups []device.RunConfigItem) error {
+	for _, rule := range cgroups {
+		// If container is running privileged and cgroup rule is for devices cgroup, apply the requested rule.
+		if c.isCurrentlyPrivileged() && !c.state.OS.RunningInUserNS && c.state.OS.CGroupDevicesController && strings.HasPrefix(rule.Key, "devices.") {
```

Comment:
What I meant with my comment is that we may soon find ourselves with calls to deviceAddCgroupRules which will not exclusively be made of `devices.` keys.

So this function should process them all and only if the prefix is devices. then skip when inside a userns or an unprivileged container.

Reply:
@stgraber I've made the change as we discussed.